### DISCUSSION TITLE
New feature ? : choose recipe and parameter destination

### DISCUSSION THREAD

**Discussion by Python-simulation** [qcha41/usit#18] [Ideas]
![image](https://github.com/autolab-project/autolab/assets/46069237/e64c22eb-cf0f-4b43-8db7-c175f5d48381)

Was trying to see if this could be a feature to replace the combobox: selectRecipe_comboBox and selectParameter_comboBox that appears in scanner when have multiple parameters and recipes.
But it seems to heavy to be user friendly. Instead of one click, it becomes two clicks for recipes and three clicks for parameters + searching among a possible long list of parameters, recipes...

Don't think I should implement it.

Alternative are:
- leave as is (combobox in scanner: currently quite cryptic)
- Move/Duplicate combobox to control panel

What do you think?

**Comment by Python-simulation** [comment]
@pyMatJ @Totorminator98 opinion please :)

**Comment by FateDecider** [comment]
I think combobox in scanner, as it is, is convenient at assembling steps. Only minor problem is - adding steps to a configuration will automatically go to the last recipe in the configuration. Not entirely a bad thing since you can drag and drop steps.

When assembling steps in the configuration, sometimes the user may not be not entirely clear which recipe the user wants the steps to be in, so I think the current setup is more user-friendly (click fewer buttons)

**Reply by Python-simulation** [reply]
I'm afraid I've misunderstood, as the purpose of the combobox is to select recipes that receive steps when added from the control panel: 
![image](https://github.com/autolab-project/autolab/assets/46069237/7ee643cb-8915-4a9d-89ab-b5d8263621b5)
However, it's true that any new step is added at the end of a recipe, but not necessarily to the last recipe (if that's where the misunderstanding comes from).

**Reply by FateDecider** [reply]
I see, that makes sense. I have always wondered what that drop-down box do. But, I think I am satisfied with what we have currently.

**Reply by pyMatJ** [reply]
Sorry, long delay. Simple addendum to make this clearer: add a label "add parameters to:" in front of the combobox? (or else, like "work on:", "modify:", "edit:"...)
I kind of like the list that you proposed though, an idea (more coding) would be to trigger it only if more recipes have been added. So if only one recipe => 1 click, otherwise yes it is a bit more tedious but fully explicit. 
IMO both work for now. 
Not in favor of moving the combobox to the control panel as is.
Thanks for the feature test though!!

**Reply by Python-simulation** [reply]
The idea was to have the menu only for multiple recipes, but that requires more coding which I didn't want to do without knowing whether the feature is useful or not. I've seen in a software a combobox split in two that allows you to click on a button and open a menu like I've done here. This would be the best way to do it, as it allows you to always have the button and optionally have the menu to change the recipe (without having to click on it each time).
In the mean time I will just add a label with 'Edit:' like you mention.

**Comment by FateDecider** [comment]
Separately, I have an important question. Is there a way I can retrieve the Parameter table in Scanner as a step in my recipe?

I was reading through the backend, and I want to save and manipulate the table obtained from ConfigManager.getParamDataFrame.

The process I have in mind is to do scanning microscopy, and I want to dynamically update the grid I want to scan, on the right of Scanner. That requires knowing the position ranges I want to scan even before starting the scan. The only way I can think of, is to directly obtain the ParamDataFrame and construct the grid out of it.

**Reply by Python-simulation** [reply]
In [v2.0b2_candidate](https://github.com/autolab-project/autolab/tree/v2.0b2_candidate) (from 2024/03/08) you can right-click on a parameter to change to 'custom values' and use any array of your choice instead of the default start/stop/step feature. But the table from getParamDataFrame cannot be modified (you may have seen my comments in gui/scanner/display.py saying that modifying a table is possible but that makes it slow to render). This could be an idea for the future if the current way of creating a custom array isn't user-friendly enough.
Currently, custom values only accept 1D arrays. ND arrays could be considered for ND scans.

**Reply by FateDecider** [reply]
Yes, that is the same as my understanding. Having the ability to readout the table of the ParamDataFrame will be quite useful for ND scans, for a variety of user-custom purposes. In general, it is still more user-friendly to add parameter values via 'Standard values'.

As for modifying the table, I am thinking of modifying a cloned copy of it via $eval, as long as I can readout the table.

**Reply by Python-simulation** [reply]
I forgot about this discussion and I see now that this could be implemented now that variables are created in a scan for each parameter and step. Currently, the most recent 'parameter' value of a parameter array from a recipe (either regular start, stop, step or log or custom) is accessible in the recipe using $eval:parameter(). For example using $eval:parameter() as step for a parameter defined as [1,2,3] will gives 1 then 2 then 3. I was thinking that in addition to creating a Variable for the parameter value, it could be useful to have a Variable for the array directly (in this example [1,2,3]). Only 'problem' is which name to use for it as the parameter name is already used for the current value and that it is not safe to assume the name <param_name>_array is usable (This comment is mostly me talking to myself).

**Comment by Python-simulation** [comment]
Choose recipe and parameter destination implemented in 58bd62b
